hugr/src/types/type_param.rs

```diff
@@ -217,6 +218,17 @@ impl TypeArg {
             TypeParam::Type { b } => TypeArg::Type {
                 ty: Type::new_var_use(idx, b),
             },
+            TypeParam::List { param: bx } if matches!(*bx, TypeParam::Type { .. }) => {
+                // There are two reasonable schemes for representing row variables:
+                // 1. TypeArg::Variable(idx, TypeParam::List(TypeParam::Type(typebound)))
+                // 2. TypeArg::Type(Type::new_row_var(idx, typebound))
+                // Here we prefer the latter for canonicalization, although we cannot really
+                // prevent both if users construct the TypeArg variants directly (doing so will break Eq)
```

Comment:
You mean so that `TypeArg::Variable(idx, TypeParam::List(TypeParam::Type(bound)))` is `==` to `TypeArg::Type(Type::RowVariable(idx, bound))` ?

Yes, we could, that is fair I guess. I would really love to stick to a canonical representation, but that is hard without going through the type system (as per branch `feat/row_typevar2` where we had `enum RowVarOrType`, or `feat/row_typevar3` where TypeArg could contain a `Type<false>` thus ruling out the RHS of the `==` example above)

I guess we could also go for the LHS of the `==` by making `impl From<Type> for TypeArg` more complex.

Reply:
Hmmmm, so I was thinking that since `TypeArg` is `pub enum`, they could construct anything. *However* `TypeArg::Variable` contains a `pub struct TypeArgVariable` and in fact this *cannot* be publically constructed - it has private members and no `new`.

Hence, the only way to construct a `TypeArg::Variable` from outside the `types` module is by `new_var_use` (or `new_row_var_use`), which means we *can* control canonicalization. Which is to say, I think the comment here is bogus.

Do you agree?

However, there is still the possibility of canonicalizing in the other direction (no `Type::RowVariable` inside `TypeArg::Type`, use `TypeArgVariable` instead). I guess I could experiment with a switcheroo commit....